File: python/packages/botas/src/botas/meter_provider.py
"""Lazy OpenTelemetry meter provider for botas metrics instrumentation.

Returns ``None`` when ``opentelemetry-api`` is not installed — all metric
operations become no-ops and the library works without telemetry.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from opentelemetry.metrics import Counter, Histogram, Meter


class BotasMetrics:
    """Pre-created metric instruments for botas."""

    def __init__(self, meter: Meter) -> None:
        self.activities_received: Counter = meter.create_counter(
            "botas.activities.received",
            description="Total activities received by the bot",
            unit="{activity}",
        )
        self.turn_duration: Histogram = meter.create_histogram(
            "botas.turn.duration",
            description="Duration of turn processing",
            unit="ms",
        )
        self.handler_errors: Counter = meter.create_counter(
            "botas.handler.errors",
            description="Total handler errors",
            unit="{error}",
        )
        self.middleware_duration: Histogram = meter.create_histogram(
            "botas.middleware.duration",
            description="Duration of individual middleware execution",
            unit="ms",
        )
        self.outbound_calls: Counter = meter.create_counter(
            "botas.outbound.calls",
            description="Total outbound API calls to Bot Service",
            unit="{call}",
        )
        self.outbound_errors: Counter = meter.create_counter(
            "botas.outbound.errors",
            description="Total outbound API call errors",
            unit="{error}",
        )


_metrics: Optional[BotasMetrics] = None
_initialized: bool = False


def get_metrics() -> Optional[BotasMetrics]:
    """Return pre-created metric instruments, or ``None`` if unavailable."""
    global _metrics, _initialized
    if _initialized:
        return _metrics
    _initialized = True
    try:
        from opentelemetry import metrics

        from botas._version import __version__

        meter = metrics.get_meter("botas", __version__)
        _metrics = BotasMetrics(meter)
    except ImportError:
        _metrics = None
    return _metrics

File: node/packages/botas-core/src/meter-provider.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import type { Counter, Histogram, Meter } from '@opentelemetry/api'
import { createRequire } from 'node:module'

const require = createRequire(import.meta.url)

let _meter: Meter | null | undefined // undefined = not yet initialized

/** @internal Reset the cached meter — for testing only. */
export function resetMeter (): void {
  _meter = undefined
  _metrics = null
}

export function getMeter (): Meter | null {
  if (_meter !== undefined) return _meter
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const api = require('@opentelemetry/api') as typeof import('@opentelemetry/api')
    const pkg = require('../package.json') as { version: string }
    _meter = api.metrics.getMeter('botas', pkg.version)
  } catch {
    _meter = null
  }
  return _meter
}

/** Pre-created metric instruments for botas. */
export interface BotasMetrics {
  /** Counter: total activities received, tagged by activity.type */
  activitiesReceived: Counter
  /** Histogram: turn processing duration in milliseconds */
  turnDuration: Histogram
  /** Counter: handler errors, tagged by activity.type */
  handlerErrors: Counter
  /** Histogram: middleware execution duration in milliseconds */
  middlewareDuration: Histogram
  /** Counter: outbound API calls to Bot Service */
  outboundApiCalls: Counter
  /** Counter: outbound API errors */
  outboundApiErrors: Counter
}

let _metrics: BotasMetrics | null = null

/**
 * Returns pre-created metric instruments, or `null` when `@opentelemetry/api` is unavailable.
 */
export function getMetrics (): BotasMetrics | null {
  if (_metrics !== null) return _metrics

  const meter = getMeter()
  if (!meter) return null

  _metrics = {
    activitiesReceived: meter.createCounter('botas.activities.received', {
      description: 'Total activities received by the bot',
      unit: '{activity}',
    }),
    turnDuration: meter.createHistogram('botas.turn.duration', {
      description: 'Duration of turn processing',
      unit: 'ms',
    }),
    handlerErrors: meter.createCounter('botas.handler.errors', {
      description: 'Total handler errors',
      unit: '{error}',
    }),
    middlewareDuration: meter.createHistogram('botas.middleware.duration', {
      description: 'Duration of individual middleware execution',
      unit: 'ms',
    }),
    outboundApiCalls: meter.createCounter('botas.outbound.calls', {
      description: 'Total outbound API calls to Bot Service',
      unit: '{call}',
    }),
    outboundApiErrors: meter.createCounter('botas.outbound.errors', {
      description: 'Total outbound API call errors',
      unit: '{error}',
    }),
  }

  return _metrics
}
